docs/custom-plugins.md

```diff
@@ -0,0 +1,143 @@
+    warnings.flush()
+
+    
+    for (const optionsEntry of options) {
+      const bundle = await rollup.rollup(optionsEntry)
+      await Promise.all(optionsEntry.output.map(bundle.write))
+    }
+  }
+}
+
+exports.tasks = [Rollup]
+```
+
+Then, in the plugin's `.toolkitrc.yml`, list this task as the default task to run on the hooks you need:
+
+```yml
+hooks:
+  'build:local': Rollup
+  'build:ci': Rollup
+  'build:remote: Rollup
+```
+
+You should install `@dotcom-tool-kit/types` and the tooling you're implementing as `devDependencies` of your repo (e.g. `npm install --save-dev @dotcom-tool-kit/types rollup`).
+
+## Implementing a new hook
+
+A Hook defines an abstract label to run tasks with, as well as managing where in other project configuration it's run from. For example, the built-in `circleci` plugin defines a `build:ci` hook, which lets tasks like Rollup run in CI, and it specifies that `build:ci` should be run by a CircleCI job and automatically manages the configuration in `.circleci/config.yml` to run that job.
+
+This abstraction lets us write different plugins for defining tasks to be run and where they should be run from, while maintaining the link between them. We've already seen that `build:ci` could be running Rollup, or Webpack, or any other task; in addition, `build:ci` itself could be defined by a different plugin, such as a Github Actions plugin, that would automatically manage configuration in `.github/workflows`.
+
+The automatic configuration management is implemented by `Hook` subclasses. These define a `check` method that should return `true` if the hook is correctly installed in the repository or `false` if it needs installing, and an `install` method to actually perform the installation. Every time Tool Kit runs, it checks that every hook is installed in your repo, and if any aren't, it exits with an error (to ensure the repo is always consistent with what it expects). You can then run `dotcom-tool-kit --install` to run the installation of every hook that isn't installed.
+
+If you find yourself asking a question like "how do I run a Tool Kit task from a different npm script", **you should implement a hook** to allow Tool Kit to automatically manage that configuration for any new repos using your plugin, rather than expecting new users to add that configuration themselves when installing the plugin.
+
+Hooks have a loose naming convention of `category:environment`. This is only meant for humans to be able to intuitively understand which hooks are related; it's not required by the Tool Kit core itself.
+
+Let's say you want to run some task on the npm `prepare` script (which automatically runs after `npm install` and before `npm publish`). We'll call that hook `prepare:local`. Create a subclass of the `Hook` class from `@dotcom-tool-kit/types`, implement the `check` and `install` methods, and export a `hooks` object to map it to the name we're giving it. Your `index.js` might include:
```

Comment:
I think it may be useful here to explain that a hook would be in a subdirectory in `toolkit` with an appropriate name, e.g. prepare(?). Also that the index file can contain the hook alone, or be in conjunction with a task. Or actually maybe do we want it do be only in conjunction with a task?